File: libs/language/src/lib/utils/NodeTransformer.ts
import { BlueNode } from '../model';

/**
 * Utility class for transforming BlueNodes
 */
export class NodeTransformer {
  /**
   * Transforms a node and all its child nodes using the provided transformer function
   * @param node - The node to transform
   * @param transformer - The transformer function to apply to each node
   * @returns The transformed node
   */
  static transform(
    node: BlueNode,
    transformer: (node: BlueNode) => BlueNode
  ): BlueNode {
    // First apply the transformer to this node
    const transformedNode = transformer(node.clone());

    const type = transformedNode.getType();
    if (type !== undefined) {
      transformedNode.setType(NodeTransformer.transform(type, transformer));
    }

    const itemType = transformedNode.getItemType();
    if (itemType !== undefined) {
      transformedNode.setItemType(
        NodeTransformer.transform(itemType, transformer)
      );
    }

    const keyType = transformedNode.getKeyType();
    if (keyType !== undefined) {
      transformedNode.setKeyType(
        NodeTransformer.transform(keyType, transformer)
      );
    }

    const valueType = transformedNode.getValueType();
    if (valueType !== undefined) {
      transformedNode.setValueType(
        NodeTransformer.transform(valueType, transformer)
      );
    }

    const items = transformedNode.getItems();
    if (items !== undefined) {
      const transformedItems = items.map((item) =>
        NodeTransformer.transform(item, transformer)
      );
      transformedNode.setItems(transformedItems);
    }

    const properties = transformedNode.getProperties();
    if (properties !== undefined) {
      const transformedProperties = Object.keys(properties).reduce(
        (acc, key) => {
          acc[key] = NodeTransformer.transform(properties[key], transformer);
          return acc;
        },
        {} as Record<string, BlueNode>
      );
      transformedNode.setProperties(transformedProperties);
    }

    const contracts = transformedNode.getContracts();
    if (contracts !== undefined) {
      const transformedContracts = Object.keys(contracts).reduce((acc, key) => {
        acc[key] = NodeTransformer.transform(contracts[key], transformer);
        return acc;
      }, {} as Record<string, BlueNode>);
      transformedNode.setContracts(transformedContracts);
    }

    return transformedNode;
  }
}

File: references/java-sdk/src/main/java/blue/language/utils/NodeTransformer.java
package blue.language.utils;

import blue.language.model.Node;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class NodeTransformer {
    public static Node transform(Node node, Function<Node, Node> nodeTransformer) {
        if (node == null) {
            return null;
        }

        Node transformedNode = nodeTransformer.apply(node.clone());

        if (transformedNode.getType() != null) {
            transformedNode.type(transform(transformedNode.getType(), nodeTransformer));
        }

        if (transformedNode.getItemType() != null) {
            transformedNode.itemType(transform(transformedNode.getItemType(), nodeTransformer));
        }

        if (transformedNode.getKeyType() != null) {
            transformedNode.keyType(transform(transformedNode.getKeyType(), nodeTransformer));
        }

        if (transformedNode.getValueType() != null) {
            transformedNode.valueType(transform(transformedNode.getValueType(), nodeTransformer));
        }

        if (transformedNode.getItems() != null) {
            List<Node> transformedItems = transformedNode.getItems().stream()
                    .map(item -> transform(item, nodeTransformer))
                    .collect(Collectors.toList());
            transformedNode.items(transformedItems);
        }

        if (transformedNode.getProperties() != null) {
            Map<String, Node> transformedProperties = new HashMap<>();
            for (Map.Entry<String, Node> entry : transformedNode.getProperties().entrySet()) {
                transformedProperties.put(entry.getKey(), transform(entry.getValue(), nodeTransformer));
            }
            transformedNode.properties(transformedProperties);
        }

        return transformedNode;
    }
}